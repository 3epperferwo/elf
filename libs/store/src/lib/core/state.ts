type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type State<State, Config> = { state: State; config: Config };
export type EmptyConfig = Record<string, any>;

export function createState<State extends any[]>(
  ...state: State
): { state: Merge<State, 'state'>; config: Merge<State, 'config'> } {
  return state.reduce(
    (acc, current) => {
      acc.config = {
        ...acc.config,
        ...current.config,
      };

      acc.state = {
        ...acc.state,
        ...current.state,
      };

      return acc;
    },
    { config: {}, state: {} }
  );
}
