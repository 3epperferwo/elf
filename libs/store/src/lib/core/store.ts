import { BehaviorSubject, Observable } from 'rxjs';
import { produce } from 'immer';

export type Reducer<State> = (state: State, store: Store) => State;

export interface StoreDef<State = any> {
  name: string;
  state: State;
  config: any;
}

export class Store<SDef extends StoreDef = any, State = SDef['state']> extends BehaviorSubject<State> {
  private currentValue: State;

  constructor(public storeDef: SDef) {
    super(storeDef.state);
    this.currentValue = this.getValue();
  }

  get state() {
    return this.currentValue;
  }

  get name() {
    return this.storeDef.name;
  }

  getConfig<Config>(): Config {
    return this.storeDef.config;
  }

  reduce(...reducers: Array<Reducer<State>>) {
    const nextState = reducers.reduce((value, reducer) => {
      value = reducer(value, this);

      return value;
    }, this.currentValue);

    if(nextState !== this.currentValue) {
      super.next(nextState);
    }

  }

  combine<R extends Observables>(observables: R): Observable<ReturnTypes<R>> {
    let hasChange = true;
    const buffer: any = [];

    return new Observable(observer => {
      observables.forEach((query, i) => observer.add(
        query.subscribe(value => {
          buffer[i] = value;
          hasChange = true;
        })));

      return this.subscribe({
        next() {
          if(hasChange) {
            observer.next(buffer);
            hasChange = false;
          }
        },
        error(e) {
          observer.error(e);
        },
        complete() {
          observer.complete();
        }
      });
    });

  }
}

type ReturnTypes<T extends Observable<any>[]> = { [P in keyof T]: T[P] extends Observable<infer R> ? R : never };
type Observables = [Observable<any>] | Observable<any>[];
