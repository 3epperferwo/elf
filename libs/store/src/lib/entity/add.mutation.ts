import { BaseEntityOptions, DefaultEntitiesRef, defaultEntitiesRef, EntitiesRecord, EntitiesRef, getEntityType, getIdKey, getIdType } from './entity.state';
import { coerceArray } from '../core/utils';
import { OrArray } from '../core/types';
import { Reducer, Store } from '../core/store';

/**
 *
 * Add entities to the store
 *
 * store.reduce(addEntity(entity))
 *
 * store.reduce(addEntity([entity, entity]))
 *
 * store.reduce(addEntity([entity, entity]), { overwrite: true })
 *
 * store.reduce(addEntity([entity, entity]), { prepend: true })
 */
export function addEntity<S extends EntitiesRecord, Ref extends EntitiesRef = DefaultEntitiesRef>(
  entities: OrArray<getEntityType<S, Ref>>,
  options: {
    prepend?: boolean;
    overwrite?: boolean;
  } & BaseEntityOptions<Ref> = {}
): Reducer<S> {
  return function reducer(state: S, store: Store) {
    const { overwrite = false, prepend = false, ref = defaultEntitiesRef } = options;

    const { entitiesKey, idsKey } = ref!;

    const toObject = {} as Record<getIdType<S, Ref>, getEntityType<S, Ref>>;
    const ids = [] as getIdType<S, Ref>;
    const idKey = getIdKey<getIdType<S, Ref>>(store);

    for(const entity of coerceArray(entities)) {
      const id = entity[idKey];
      if(!overwrite && id in state[entitiesKey]) continue;

      ids.push(id);
      toObject[id] = entity;
    }

    if(!ids.length) return state;

    return {
      ...state,
      [entitiesKey]: overwrite ? toObject : { ...state[entitiesKey], ...toObject },
      [idsKey]: overwrite ? ids : prepend ? [...ids, ...state[idsKey]] : [...state[idsKey], ...ids]
    };
  };
}
