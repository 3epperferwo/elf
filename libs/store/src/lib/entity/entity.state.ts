import { Store } from '@eleanor/store';

export const defaultEntitiesKey = '$entities' as const;
export const defaultIdsKey = '$ids' as const;

export class EntitiesRef<EntitiesKey extends string = any, IdsKey extends string = any> {
  entitiesKey: EntitiesKey;
  idsKey: IdsKey;

  constructor(private config: { entitiesKey: EntitiesKey, idsKey: IdsKey }) {
    this.entitiesKey = config.entitiesKey ?? defaultEntitiesKey;
    this.idsKey = config.idsKey ?? defaultIdsKey;
  }
}

export const defaultEntitiesRef = new EntitiesRef({ entitiesKey: defaultEntitiesKey, idsKey: defaultIdsKey });

export const entitiesUIRef = new EntitiesRef({
  entitiesKey: '$UIEntities',
  idsKey: '$UIIds'
});

export function withEntitiesFactory<S extends EntitiesRecord,
  Ref extends EntitiesRef,
  >(ref: Ref) {
  return {
    [ref.entitiesKey]: {},
    [ref.idsKey]: []
  } as unknown as {
    [P in Ref['entitiesKey'] | Ref['idsKey']]: S[P]
  };
}

export function withEntities<EntityType, IdType extends PropertyKey>(config: Partial<Config> = {}) {
  return {
    state: withEntitiesFactory<EntityState<EntityType, IdType>, typeof defaultEntitiesRef>(defaultEntitiesRef),
    config: {
      idKey: config?.idKey ?? 'id'
    }
  };
}

export function withUIEntities<EntityType, IdType extends PropertyKey>(config: Partial<Config> = {}) {
  return {
    state: withEntitiesFactory<UIEntityState<EntityType, IdType>, typeof entitiesUIRef>(entitiesUIRef),
    config: {
      idKey: config?.idKey ?? 'id'
    }
  };
}

export function getIdKey<T>(store: Store): T {
  return store.getConfig<Config>().idKey as unknown as T;
}

export interface EntityState<EntityType = any, IdType extends PropertyKey = any> {
  $entities: Record<IdType, EntityType>;
  $ids: Array<IdType>;
}

export interface UIEntityState<EntityType = any,
  IdType extends PropertyKey = any> {
  $UIEntities: Record<IdType, EntityType>;
  $UIIds: Array<IdType>;
}

interface Config {
  idKey: string;
}

export type getEntityType<S extends EntitiesRecord, Ref extends EntitiesRef> = S[Ref['entitiesKey']][0];
export type getIdType<S extends EntitiesRecord, Ref extends EntitiesRef> = S[Ref['idsKey']][0];
export type ItemPredicate<Item> = (item: Item, index?: number) => boolean;
export type EntitiesRecord = Record<any, any>;
export type DefaultEntitiesRef = typeof defaultEntitiesRef;

export interface BaseEntityOptions<Ref extends EntitiesRef> {
  ref?: Ref;
}
