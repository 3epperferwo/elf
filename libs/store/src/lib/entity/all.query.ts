import { EntityState, getEntityType } from './entity.state';
import { MonoTypeOperatorFunction, OperatorFunction, pipe } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';

export function untilEntitiesChanges<T extends EntityState>(): MonoTypeOperatorFunction<T> {
  return distinctUntilChanged(
    (prev, current) => prev.$entities === current.$entities
  );
}

export function selectAll<S extends EntityState>(): OperatorFunction<S, getEntityType<S>[]> {
  return pipe(
    untilEntitiesChanges(),
    map(state => state.$ids.map((id) => state.$entities[id]))
  );
}
