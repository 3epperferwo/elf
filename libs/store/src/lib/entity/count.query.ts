import { OperatorFunction, pipe } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { EntityState, getEntityType, ItemPredicate } from './entity.state';
import { untilEntitiesChanges } from './all.query';
import { select } from '../core/queries';

/**
 * Select the store's entity collection length
 *
 * @example
 *
 * store.pipe(selectCount())
 *
 */
export function selectCount<T extends EntityState>(): OperatorFunction<T, number> {
  return select(state => state.$ids.length);
}

/**
 * Select the store's entity collection length
 *
 * @example
 *
 * store.pipe(selectCount(entity => entity.completed))
 *
 */
export function selectCountByPredicate<T extends EntityState>(predicate: ItemPredicate<getEntityType<T>>): OperatorFunction<T, number> {
  return pipe(
    untilEntitiesChanges(),
    map(state => state.$ids.filter((id) => predicate(state.$entities[id])).length),
    distinctUntilChanged()
  );
}
