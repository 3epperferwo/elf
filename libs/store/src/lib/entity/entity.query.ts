import { OperatorFunction } from 'rxjs';
import { select } from '../core/queries';
import { isString, isUndefined } from '../core/utils';
import { BaseEntityOptions, defaultEntitiesRef, DefaultEntitiesRef, EntitiesRecord, EntitiesRef, getEntityType, getIdType } from './entity.state';

interface Options extends BaseEntityOptions<any> {
  pluck?: string | ((entity: unknown) => any);
}

/**
 * Select entity from the store
 *
 * @example
 *
 * store.pipe(selectEntity(id, { pluck: 'title' })
 *
 */
export function selectEntity<S extends EntitiesRecord, K extends keyof getEntityType<S, Ref>, Ref extends EntitiesRef = DefaultEntitiesRef>(
  id: getIdType<S, Ref>,
  options: { pluck: K } & BaseEntityOptions<Ref>
): OperatorFunction<S, getEntityType<S, Ref>[K]>;

/**
 * Select entity from the store
 *
 * @example
 *
 * store.pipe(selectEntity(id, { pluck: e => e.title })
 *
 */
export function selectEntity<S extends EntitiesRecord, R, Ref extends EntitiesRef = DefaultEntitiesRef>(
  id: getIdType<S, Ref>,
  options: { pluck: (entity: getEntityType<S, Ref>) => R } & BaseEntityOptions<Ref>
): OperatorFunction<S, R>;

/**
 * Select entity from the store
 *
 * @example
 *
 * store.pipe(selectEntity(id)
 *
 */
export function selectEntity<S extends EntitiesRecord, Ref extends EntitiesRef = DefaultEntitiesRef>(
  id: getIdType<S, Ref>,
  options?: BaseEntityOptions<Ref>
): OperatorFunction<S, getEntityType<S, Ref>>;

export function selectEntity<S extends EntitiesRecord, R>(id: any, options: Options = {}) {
  const { ref: { entitiesKey, pluck } = defaultEntitiesRef } = options;

  return select<S, R>(state => getEntity(state[entitiesKey], id, pluck));
}

export function getEntity(
  entities: EntitiesRecord,
  id: string | number,
  pluck: Options['pluck']
) {
  const entity = entities[id];

  if(isUndefined(entity)) {
    return undefined;
  }

  if(!pluck) {
    return entity;
  }

  if(isString(pluck)) {
    return entity[pluck];
  }

  return pluck(entity);
}
