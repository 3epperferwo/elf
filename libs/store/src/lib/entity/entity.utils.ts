import { EntitiesRecord, EntitiesRef, getEntityType, getIdType } from './entity.state';
import { coerceArray } from '../core/utils';

export function buildEntities<S extends EntitiesRecord, Ref extends EntitiesRef>(ref: Ref, entities: getEntityType<S, Ref>[], idKey: string):
  { ids: getIdType<S, Ref>[], asObject: getEntityType<S, Ref> } {
  const asObject = {} as Record<getIdType<S, Ref>, getEntityType<S, Ref>>;
  const ids = [] as getIdType<S, Ref>;

  for(const entity of coerceArray(entities)) {
    const id: getIdType<S, Ref> = entity[idKey];

    ids.push(id);
    asObject[id] = entity;
  }

  return {
    ids,
   asObject
  } as unknown as { ids: getIdType<S, Ref>[], asObject: getEntityType<S, Ref> };
}
