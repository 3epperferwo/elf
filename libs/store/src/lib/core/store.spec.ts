import { createTodo, createUITodo, Todo } from '../mocks/stores.mock';
import {
  addEntities,
  selectAll,
  updateEntities,
  withEntities,
  withUIEntities,
} from '../entities';
import { withProps } from '../props/props.state';
import { Store } from './store';
import { entitiesUIRef } from '../entities/entity.state';
import { createState } from './state';

type UIEntity = { id: number; open: boolean };

describe('store', () => {
  describe('combine', () => {
    const { state, config } = createState(
      withEntities<Todo>(),
      withUIEntities<UIEntity>(),
      withProps<{ filter: string }>({ filter: '' })
    );

    const store = new Store({ state, name: 'todos', config });

    it('should fire only once', () => {
      const spy = jest.fn();

      store
        .combine([
          store.pipe(selectAll()),
          store.pipe(selectAll({ ref: entitiesUIRef })),
        ])
        .subscribe(spy);

      expect(spy).toHaveBeenCalledTimes(1);

      expect(store.state).toMatchSnapshot();

      store.reduce(
        addEntities(createTodo(1)),
        addEntities(createUITodo(1), { ref: entitiesUIRef })
      );

      expect(spy).toHaveBeenCalledTimes(2);
      expect(store.state).toMatchSnapshot();

      store.reduce((state) => ({ ...state, filter: 'foo' }));

      // Update non related value should not call `next`
      expect(spy).toHaveBeenCalledTimes(2);

      store.reduce(updateEntities(1, { title: 'foo' }), (state) => ({
        ...state,
        filter: 'hello',
      }));

      expect(store.state).toMatchSnapshot();
      expect(spy).toHaveBeenCalledTimes(3);
    });
  });
});
