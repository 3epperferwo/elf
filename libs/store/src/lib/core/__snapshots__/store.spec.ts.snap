// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store combine should fire only once 1`] = `
Object {
  "$UIEntities": Object {},
  "$UIIds": Array [],
  "$entities": Object {},
  "$ids": Array [],
  "filter": "",
}
`;

exports[`store combine should fire only once 2`] = `
Object {
  "$UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
  },
  "$UIIds": Array [
    1,
  ],
  "$entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "$ids": Array [
    1,
  ],
  "filter": "",
}
`;

exports[`store combine should fire only once 3`] = `
Object {
  "$UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
  },
  "$UIIds": Array [
    1,
  ],
  "$entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "foo",
    },
  },
  "$ids": Array [
    1,
  ],
  "filter": "hello",
}
`;
